lib/python/flame/privacy/differential_privacy.py

```
@@ -0,0 +1,89 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from flame.common.typing import ModelWeights
+from flame.common.util import (
+    MLFramework,
+    get_ml_framework_in_use,
+)
+
+
+class DifferentialPrivacy:
+    """
+    Differential Privacy class that adds noise to the trainer's update weights.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        """Initialize instance."""
+
+        try:
+            # Fedbalancer parameters
+
+            # clip threshold
+            self._clip_threshold = kwargs["clip_threshold"]
+            # noise factor for differential privacy
+            self._noise_factor = kwargs["noise_factor"]
+
+        except KeyError:
+            raise KeyError(
+                "one of the parameters among {clip_threshold, noise_factor} "
+                + "is not specified in config"
+            )
+
+        framework = get_ml_framework_in_use()
+        if framework == MLFramework.TENSORFLOW:
+            self.apply_dp_fn = self._apply_dp_tensorflow
+        elif framework == MLFramework.PYTORCH:
+            import torch as T
+
+            self.torch = T
+            self.apply_dp_fn = self._apply_dp_torch
+
+    def _apply_dp_torch(self, delta_weights: ModelWeights) -> ModelWeights:
+        """Apply dp noise to the delta_weights."""
+        delta_weights = self._apply_clip_norm_torch(delta_weights, self._clip_threshold)
+        return {
+            k: d_w + self.torch.normal(mean=0, std=self._noise_factor, size=d_w.shape)
+            for k, d_w in delta_weights.items()
+        }
+
+    # TODO: Currently, dp is only supported on PyTorch.
```

Jira ticket?

---

Unless you have quick plans to adjust this, I would say that it's better to raise an exception (NotImplementedError) in the constructor than to let it run with tensorflow with no actual effect. Is there a reason why you want to keep it this way?